imports/client/components/PuzzlePage.tsx

```diff
@@ -134,78 +134,93 @@ interface ChatHistoryProps {
+    // mistakenly save an incorrect scrollBottomTarget.  So skip one scroll event when we self-induce
+    // this callback, so we only update the target distance from bottom when the user is actually
+    // scrolling.
+    if (shouldIgnoreNextScrollEvent.current) {
+      shouldIgnoreNextScrollEvent.current = false;
+    } else {
+      saveScrollBottomTarget();
+    }
+  }, [saveScrollBottomTarget]);
+
+  const scrollToTarget = useCallback(() => {
+    if (ref.current) {
+      const rect = ref.current.getClientRects()[0];
+      const hiddenHeight = ref.current.scrollHeight - rect.height;
+      // if distanceFromBottom is hiddenHeight - scrollTop, then
+      // our desired scrollTop is hiddenHeight - distanceFromBottom
+      const scrollTopTarget = hiddenHeight - scrollBottomTarget.current;
+      if (ref.current.scrollTop !== scrollTopTarget) {
+        shouldIgnoreNextScrollEvent.current = true;
+        ref.current.scrollTop = scrollTopTarget;
+      }
     }
   }, []);
 
-  const maybeForceScrollBottom = useCallback(() => {
-    console.log('maybeForceScrollBottom called');
-    if (shouldScroll.current) {
-      forceScrollBottom();
-    }
-  }, [forceScrollBottom]);
+  const snapToBottom = useCallback(() => {
+    scrollBottomTarget.current = 0;
+    scrollToTarget();
+  }, [scrollToTarget]);
 
   useImperativeHandle(forwardedRef, () => ({
-    forceScrollBottom,
-    maybeForceScrollBottom,
+    saveScrollBottomTarget,
+    snapToBottom,
+    scrollToTarget,
   }));
 
-  const resizeHandler = useCallback(() => {
-    maybeForceScrollBottom();
-  }, [maybeForceScrollBottom]);
-
   useEffect(() => {
     // Scroll to end of chat on initial mount.
-    forceScrollBottom();
-  }, [forceScrollBottom]);
+    scrollToTarget();
+  }, [scrollToTarget]);
 
   useEffect(() => {
-    // Add resize handler
-    window.addEventListener('resize', resizeHandler);
+    // Add resize handler that scrolls to target
+    window.addEventListener('resize', scrollToTarget);
 
     return () => {
-      window.removeEventListener('resize', resizeHandler);
+      window.removeEventListener('resize', scrollToTarget);
     };
-  }, [resizeHandler]);
-
-  useEffect(() => {
-    // Whenever we rerender, check if we should be scrolling to the bottom.
-    maybeForceScrollBottom();
-  });
+  }, [scrollToTarget]);
+
+  useLayoutEffect(() => {
+    // Whenever we rerender due to new messages arriving, make our
+    // distance-from-bottom match the previous one, if it's larger than some
+    // small fudge factor.  But if the user has actually scrolled into the backlog,
+    // don't move the backlog while they're reading it -- instead, assume they want
+    // to see the same messages in the same position, and adapt the target bottom
+    // distance instead.
+    if (scrollBottomTarget.current > 10) {
+      saveScrollBottomTarget();
+    } else {
+      scrollToTarget();
```

Comment:
I find this behavior a little odd. If you're scrolled to the last line, but it's slightly cut off, it will continue to be slightly cut off as new messages arrive. Since the assumption within this margin is that the user intended to be at the bottom, I think we should snap all the way to the bottom when a new message arrives.
```suggestion
      snapToBottom();
```